### ISSUE TITLE
Reduce huge heap allocations during Regex instantiation

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/corefx#30507]
While working on https://github.com/dotnet/corefx/pull/30474 and looking at https://benchmarksgame-team.pages.debian.net/benchmarksgame/performance/regexredux.html I noticed immensely high heap allocations when constructing Regex objects (see benchmark tables). Looking at the code it speaks for itself that the most allocations happen during the parsing of the pattern. This should and can easily be reduced by using ref structs, Span, ArrayPool and stack allocation where possible. One example is the RegexTree which is documented as "not space efficient but convenient".

cc @danmosemsft @stephentoub

**Comment by MarcoRossignoli** [comment]
>  I noticed immensely high heap allocations when constructing Regex objects (see benchmark tables).

The benchmarks are on codebase(i check [here](https://github.com/dotnet/corefx/tree/master/src/System.Text.RegularExpressions/tests/Performance) but i didn't found)?If not could you share(it could be useful as a starting point)?

**Comment by ViktorHofer** [comment]
https://benchmarksgame-team.pages.debian.net/benchmarksgame/performance/regexredux.html

**Comment by MarcoRossignoli** [comment]
@ViktorHofer @adamsitnik i followed this guide https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/benchmarking.md.
I compiled corefx and create a new project on my [lab repo](https://github.com/MarcoRossignoli/marcorossignoli.github.io/tree/master/corefx/src/Lab/RegExPerf).
But when run i get:
` error NU1101: Unable to find package runtime.win-x64.Microsoft.NETCore.App. No packages exist with this id in source(s): CoreClrNuGetFeed, local clr`
In my local packages folder there is `runtime.win-x64.Microsoft.Private.CoreFx.NETCoreApp.4.6.0-preview1-26721-0.symbols.nupkg` and on autogenerated config there is correct include `<PackageReference Include="runtime.win-x64.Microsoft.Private.CoreFx.NETCoreApp" Version="4.6.0-preview1-26721-0" />`
My suspect are that if i open package lib are under netcoreapp3.0 and runtime try to find netcoreapp2.1.
I know there are other way to do perf analysis but i'd like to use this elegant way if possible, maybe i'm missing something.

**Comment by ViktorHofer** [comment]
I don't know if we already mapped netcoreapp to netcoreapp3.0. @weshaggard should know.
The error you are hitting is the same that I hit when I tested that feature for @adamsitnik. I wasn't able to work around it and hadn't anymore time to test it thoroughly.
@adamsitnik we should try to stabilize that feature, sorry that I didn't get back to you. If you want I can block some time this week to try it again.

**Comment by ViktorHofer** [comment]
@MarcoRossignoli I'm currently using a different testing paradigm. I guess you are currently blocked :/

**Comment by weshaggard** [comment]
Yes the latest version of of these packages are putting assets in netcoreapp3.0 so you need to target that.

**Comment by MarcoRossignoli** [comment]
> @MarcoRossignoli I'm currently using a different testing paradigm. I guess you are currently blocked :/

@ViktorHofer  don't worry...i'll try other way. Can you describe your way, i'd like to know new methods?

**Comment by MarcoRossignoli** [comment]
> Yes the latest version of of these packages are putting assets in netcoreapp3.0 so you need to target that.

@weshaggard but if download last runtime and try to compile with netcoreapp3.0 i get SDK doesn't support 3.0 but only 2.2, where is sdk that support netcoreapp3.0?
My idea was to change default dotnet.exe(on benchmark config) with nightly sdk build support 3.0 if possible.

**Comment by weshaggard** [comment]
It doesn't exist yet. For now you can do https://github.com/dotnet/corefx/blob/master/pkg/test/frameworkSettings/netcoreapp3.0/settings.targets#L5 in your project to allow the SDK build netcoreapp3.0 applications.

**Comment by MarcoRossignoli** [comment]
@adamsitnik thank's!
I'll try new way and i'll let you know!

**Comment by AraHaan** [comment]
@ViktorHofer is this still available to fix? or has it been fixed already? If not I would like to take this issue.

**Comment by ViktorHofer** [comment]
If my memory serves me right I reduced allocations by >10%. We definitely could still do more.

**Comment by AraHaan** [comment]
Alright then. I could try to do some experimentations with it. However if it results in better performance it might not be fixed until after october.

**Comment by danmosemsft** [comment]
@pgovind if you want to spend time on regex, it is likely there are worthwhile changes possible here.
